### ISSUE TITLE
perf: a compiled hook loads the entire CLI to decide allow/deny — 544ms per spawn

### ISSUE THREAD

**Issue by zernie** [zernie/agent-lint#216]
`vigiles hook-runtime run-program` is on the hot path of every gated tool call. Measured on this machine, Node 22:

| | per invocation |
|---|---|
| bare `node -e ''` | **33 ms** |
| `require("dist/hook.js")` — the hook vocabulary alone | **137 ms** |
| `require("dist/cli.js")` — what the runtime actually loads | **492 ms** |
| a real hook run (`run-program` over `examples/harness/safe-bash-guard.mjs`) | **544 ms** |

`dist/cli.js` is 357 KB and carries `compile`, `lint`, `audit`, `scan`, `eval`, the report template loader and everything else. None of it is reachable from the decision a gate makes. So roughly **490 ms of the 544 ms is spent loading code the hook never calls**, on every single spawn.

## Shape of the fix

A dedicated minimal entrypoint for the runtime path, importing only what a decision needs (the hook vocabulary, the shell parser, the protocol encode), with the CLI's own verbs left behind a lazy import. The floor is whatever `dist/hook.js` costs — 137 ms today, itself worth a look, since a decision is pure computation over a small JSON payload.

Worth measuring alongside: how much of the remaining 137 ms is the shell parser, which a gate that never calls `runs()`/`touches()` does not need either.

## Relation to emitting a native `if:`

An earlier revision of this issue argued the fix removes the reason to emit Claude Code's native `if:` at all. That framing was wrong and is withdrawn — it pre-judged a decision this measurement does not settle. The two are complementary: a native `if:` reduces how OFTEN the hook spawns, this issue reduces what each spawn COSTS. Both are worth having.

For the record, since it is the reason this issue was opened in the first place: a native `if:` is only safe when it is emitted as a deliberate WIDENING of the typed predicate, never as its natural translation. Measured on Claude Code 2.1.263 (`tools/measure-if-matcher-forms.mjs`, and the widening cases alongside it):

| emitted `if:` | `git push --force` | `/usr/bin/git push --force` | `ls -la` |
|---|---|---|---|
| `Bash(git push *--force*)` — the natural translation | fires | **does not fire** | does not fire |
| `Bash(*git*)` — widened to the program name | fires | fires | does not fire |
| `Bash(*)` | fires | fires | fires |

The natural translation is strictly narrower than the body it guards, so it silently disables the guard on an absolute program head. Widening to the program name alone keeps every case the body would catch while still filtering unrelated commands. Any emission must therefore be per-primitive and carry a test that the emitted pattern is provably a superset of the predicate — `touches()` and `pipesToShell()` have not been measured and may have no useful widening at all.

## How to reproduce

```bash
npm run build
printf '{"hook_event_name":"PreToolUse","tool_name":"Bash","tool_input":{"command":"ls -la"}}' > /tmp/ev.json
time (for i in $(seq 10); do node dist/cli.js hook-runtime run-program examples/harness/safe-bash-guard.mjs < /tmp/ev.json >/dev/null; done)
time (for i in $(seq 10); do node -e 'require("./dist/cli.js")' >/dev/null; done)
time (for i in $(seq 10); do node -e 'require("./dist/hook.js")' >/dev/null; done)
```

A regression test should assert the runtime entrypoint does not pull the CLI barrel — a module-graph assertion rather than a timing one, since a timing threshold is flaky on shared CI runners and would be the first thing quarantined.
